pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+    Trotterize the Hamiltonian and with a single step
+
+    .. math:
+
+        U = \prod_i \exp{-it O_i}
+
+    Then this unitary is appended to the current circuit.
+
+    Args:
+        tape: qml.QuantumTape of circuit
+        hamiltonian: qml.Hamiltonian object
+        t: time evolution size
+
+    Returns:
+
+    """
+    for obj in tape.operations:
+        qml.apply(obj)
+    qml.templates.ApproxTimeEvolution(hamiltonian, t, 1)
+
+    for obj in tape.measurements:
+        qml.apply(obj)
+
+
+class LieGradientOptimizer:
+    r"""Exact Lie gradient optimizer"""
+
+    # pylint: disable=too-few-public-methods
+    pl_paulis = [
+        qml.Identity,
+        qml.PauliX,
+        qml.PauliY,
+        qml.PauliZ,
+    ]
+    pauli_int_to_str = {0: "I", 1: "X", 2: "Y", 3: "Z"}
+    pauli_str_to_int = {"I": 0, "X": 1, "Y": 2, "Z": 3}
+
+    def __init__(self, circuit, hamiltonian, stepsize=0.01, **kwargs):
+        r"""
+        Base class for other gradient-descent-based optimizers.
+
+        A step of the Lie gradient iterates the Lie gradient flow on :math:`\text{SU}(2^N)`.
+        The function to be minimized is :math:`f(U) = \text{Tr}(U \rho_0 U^\dag H)` given a
+        Hamiltonian :math:`H` and initial state :math:`\rho_0`
+
+        .. math::
+
+            U^{(t+1)} = \exp{\epsilon \text{grad}f(U^{(t)}}) U^{(t)}
+
+        where :math:`\epsilon` is a user-defined hyperparameter corresponding to step size.
+
+        The Lie gradient is given by.
+
+        .. math::
+
+             \text{grad}f(U^{(t)}}) = -[U \rho U^\dag, H]
+
+        Subsequent steps of this optimizer will append a Trotterized version of the exact Lie
+        gradient and grow the circuit.
+
+        Args:
+            circuit (qml.QNode): the user-defined hyperparameter :math:`\eta`
+            hamiltonian (qml.Hamiltonian): the user-defined hyperparameter :math:`\eta`
+            stepsize (float): the user-defined hyperparameter :math:`\eta`
+            **kwargs
+        **Examples:**
+
+        >>> @qml.qnode(qml.device("default.qubit", wires=2))
+        ... def quant_fun():
+        ...     qml.RX(0.1, wires=[0])
+        ...     qml.RY(0.5, wires=[1])
+        ...     qml.CNOT(wires=[0,1])
+        ...     qml.RY(0.6, wires=[0])
+        ...     return qml.state()
+        Define a Hamiltonian cost function to minimize.
+        >>> hamiltonian = qml.Hamiltonian(coeffs=[-1.]*3, observables=[qml.PauliX(0), qml.PauliZ(1), qml.PauliY(0)@qml.PauliX(1)])
+        Instatiate the optimizer with the initial circuit and the cost function. Set the stepsize
+        accordingly.
+        >>> opt = LieGradientOptimizer(circuit = quant_fun, hamiltonian=hamiltonian, stepsize=0.1)
+        Applying 10 steps gets us close the ground state of E=-2.23
+        >>> for step in range(10):
+        ...    print(step)
+        ...    state = opt.step_and_cost()
+        ...    print(np.trace(H @ np.outer(state, state.conj())))
```

Comment:
I tried running this locally, but run into an error because H is a `qml.Hamiltonian` object 🤔 Have you converted `H` into a numeric matrix here?

Reply:
oh is this just to view the cost function output? This could actually be an advantage of having the optimizer accept a single cost function, rather than an ansatz and variable - you will have a function to execute!